offchain-modules/packages/x/src/xchain/eth/contract.ts

```diff
@@ -36,49 +34,55 @@ export class EthChain {
     }
   }
 
-  async watchLockEvents(startHeight = 1, handleLogFunc) {
-    const confirmNumber = this.config.confirmNumber > 0 ? this.config.confirmNumber : 0;
-    let currentBlockNumber = await this.provider.getBlockNumber();
-    let maxConfirmedBlock = currentBlockNumber - confirmNumber;
-    let fromBlock = startHeight;
-    while (true) {
-      try {
-        if (fromBlock >= maxConfirmedBlock) {
-          while (true) {
-            currentBlockNumber = await this.provider.getBlockNumber();
-            maxConfirmedBlock = currentBlockNumber - confirmNumber;
-            if (fromBlock < maxConfirmedBlock) {
-              break;
-            }
-            await asyncSleep(5000);
-          }
-        }
-        let toBlock = fromBlock + BlockBatchSize;
-        if (toBlock > maxConfirmedBlock) {
-          toBlock = maxConfirmedBlock;
-        }
-        const logs = await this.provider.getLogs({
-          fromBlock: fromBlock,
-          address: this.bridgeContractAddr,
-          topics: [ethers.utils.id('Locked(address,address,uint256,bytes,bytes)')],
-          toBlock: toBlock,
-        });
-        logger.debug(
-          `EthChain watchLockEvents from:${fromBlock} to:${toBlock} currentBlockNumber:${currentBlockNumber} confirmNumber:${confirmNumber} logs:${logs.length}`,
-        );
-        for (const log of logs) {
-          logger.debug('log', log);
-          const parsedLog = this.iface.parseLog(log);
-          await handleLogFunc(log, parsedLog);
-        }
-        fromBlock = toBlock + 1;
-      } catch (err) {
-        logger.error('EthChain watchLockEvents error:', err);
-        await asyncSleep(3000);
+  watchLockEvents(startHeight = 1, handleLogFunc) {
+    const filter = {
+      address: this.bridgeContractAddr,
+      fromBlock: 'earliest',
+      topics: [lockTopic],
+    };
+    this.provider.resetEventsBlock(startHeight);
+    this.provider.on(filter, async (log) => {
+      const parsedLog = this.iface.parseLog(log);
+      await handleLogFunc(log, parsedLog);
+    });
+  }
+
+  async watchNewBlock(startHeight: number, handleBlockFunc) {
+    for (let height = startHeight + 1; ; ) {
+      const block = await this.provider.getBlock(height);
+      if (!block) {
+        await asyncSleep(5 * 1000);
+        continue;
       }
+      handleBlockFunc(block);
```

Comment:
`handleBlockFunc` is an async function, but there is no await here. It may have concurrent issue when we start from a earlier height.